File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.py
def solution(land):
    ScoreSum = [0, 0, 0, 0] # 첫번째 행 전의 임의의 0값으로 된 행이 있다고 가정
    
    for line in land:
        NewScoreSum = [0, 0, 0, 0] # 다음 행을 위한 값
        for idx in range(4):
            PrevScoreSum = ScoreSum[:idx] + ScoreSum[idx+1:] # 같은 라인 값은 빼주기
            NewScoreSum[idx] = line[idx] + max(PrevScoreSum)
        ScoreSum = NewScoreSum # 갱신해주기
    
    return max(ScoreSum)
    

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.java
class Solution {
    int solution(int[][] l) {
        int n = l.length;
        
        for (int i = 1; i < n; i++) {
            l[i][0] += Math.max(l[i-1][1], Math.max(l[i-1][2], l[i-1][3]));
            l[i][1] += Math.max(l[i-1][0], Math.max(l[i-1][2], l[i-1][3]));
            l[i][2] += Math.max(l[i-1][0], Math.max(l[i-1][1], l[i-1][3]));
            l[i][3] += Math.max(l[i-1][0], Math.max(l[i-1][1], l[i-1][2]));
        }
        
        int[] last = l[n - 1];
        return Math.max(Math.max(last[0], last[1]), Math.max(last[2], last[3]));
    }
}